### PR TITLE
Concurrently process deletion candidates within NEG Garbage Collector

### PR DESCRIPTION
**Pull request by gauravkghildiyal**
/assign @swetharepakula

### CODE REVIEW

File: pkg/neg/manager.go [hunk 1]
```diff
@@ -528,26 +535,68 @@ func (manager *syncerManager) garbageCollectNEGWithCRD() error {
...
+	manager.logger.V(2).Info("Count of NEGs referenced by SvcNegCR", "svcneg", klog.KObj(svcNegCR), "count", len(svcNegCR.Status.NetworkEndpointGroups))
+	var errList []error
+	shouldDeleteNegCR := true
+	deleteByZone := len(svcNegCR.Status.NetworkEndpointGroups) == 0
+
 	// deleteNegOrReportErr will attempt to delete the specified NEG resource in the cloud. If an error
 	// occurs, it will report an error as an event on the given CR. If an error does occur, false will
 	// be returned to indicate that the CR should not be deleted.
-	deleteNegOrReportErr := func(name, zone string, cr *negv1beta1.ServiceNetworkEndpointGroup) bool {
+	deleteNegOrReportErr := func(name, zone string, svcNegCR *negv1beta1.ServiceNetworkEndpointGroup) bool {
```
Comment (swetharepakula) [inline pkg/neg/manager.go:590]: nit: pull this out into its own function as we will be recreating this function once per service and there is no special context in the function itself.

File: pkg/neg/manager.go [hunk 2]
```diff
@@ -557,55 +606,50 @@ func (manager *syncerManager) garbageCollectNEGWithCRD() error {
...
-			}
+	func() {
+		manager.mu.Lock()
+		defer manager.mu.Unlock()
 
-			manager.logger.V(2).Info("Deleting NEG CR", "svcneg", klog.KObj(cr))
-			if err := deleteSvcNegCR(manager.svcNegClient, cr, manager.logger); err != nil {
-				errList = append(errList, err)
+		// Verify that the NEG is still not wanted before deleting the CR. Mitigates the possibility of the race
+		// condition mentioned above
```
Comment (swetharepakula) [inline pkg/neg/manager.go:636]: nit: since the comment is now in a different function, point to the preceding function

**Review by swetharepakula — COMMENTED** [review]
few nits, otherwise LGTM

File: pkg/neg/manager.go (hunk 1 above)
Comment (gauravkghildiyal) [reply pkg/neg/manager.go]: Sure! Done.

File: pkg/neg/manager.go (hunk 2 above)
Comment (gauravkghildiyal) [reply pkg/neg/manager.go]: Added.

**Review by swetharepakula — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**